webpages/settings/licenses.js

```diff
@@ -14,37 +15,43 @@ const vue = new Vue({
   },
 });
 
-chrome.runtime.sendMessage("getLibraryInfo", (libraryLicenses) => {
-  const licenseNameToText = {};
-  const searchParams = new URL(location.href).searchParams;
-  const libraryParam = searchParams.get("libraries");
-  if (typeof libraryParam !== "string") return;
-  const libraries = libraryParam.split(",");
-  console.log(libraryLicenses, libraries);
-  for (const library of libraries) {
-    const licenseName = libraryLicenses[library];
-    if (!licenseName) continue;
-    if (Object.prototype.hasOwnProperty.call(licenseNameToText, licenseName)) {
-      vue.libraries = [
-        ...vue.libraries,
-        {
-          name: library,
-          license: licenseNameToText[licenseName],
-        },
-      ];
-      continue;
-    }
-    chrome.runtime.sendMessage({ licenseName }, ({ licenseText }) => {
-      licenseNameToText[licenseName] = licenseText;
-      vue.libraries = [
-        ...vue.libraries,
-        {
-          name: library,
-          license: licenseText,
-        },
-      ];
-    });
-  }
-});
+setTimeout(() => {
+  const element = document.getElementById("loading");
+  if (element) element.style.display = "block";
+}, 250);
 
-document.title = chrome.i18n.getMessage("licensesTitle");
+const libraryParam = new URLSearchParams(location.search).get("libraries");
+if (typeof libraryParam !== "string") throw new Error("Invalid libraries parameter");
+const libraries = libraryParam.split(",");
+
+const libraryLicensesList = await fetch("/libraries/license-info.json").then((res) => res.json());
+
+// Load the license text of each library to be displayed
+let licenses = [];
+for (const library of libraries) {
+  if (libraryLicensesList[library]) licenses.push(libraryLicensesList[library]);
+}
+licenses = [...new Set(licenses)]; // Remove duplicate values
+const licenseNameToText = {};
+await Promise.all(
+  licenses.map((name) =>
+    fetch(`/libraries/licenses/${name}.txt`)
+      .then((res) => res.text())
+      .then((text) => (licenseNameToText[name] = text))
+  )
+);
+
+// Display the license of each library
+for (const library of libraries) {
```

Comment:
Do the libraries need to be looped though twice?

Reply:
Does it matter? The main reason there are two for-of loops is because of the sequence of operations that needs to happen, and there's a step between them - getting the text of each license - that is programmed to handle data in bulk.
https://github.com/ScratchAddons/ScratchAddons/blob/1c909ba58f8714cd4ed67bf91532db9cac8b680a/webpages/settings/licenses.js#L34-L42
This approach prevents redundant fetches (each file is only fetched once) and allows the fetches to complete asynchronously. For example, if two libraries both have the same license, that license text will be fetched once and then displayed in two different places on the page.

If you're just trying to understand the code, the procedure is this:
1. Get the URL parameter corresponding to which library licenses to display on the page and store it in the `libraries` array. (Note: this could be arbitrary, so we need to validate it)
2. Fetch and load the file that defines which license each library uses and store it in the `libraryLicensesList` variable.
3. **For each library** name, store the name of the license it uses in the `licenses` array, if that license is included in `libraryLicensesList`.
4. Remove duplicate values from `licenses`.
5. Fetch and load each license text file, of the licenses that we need to display.
6. **For each library** license to be displayed, show its license text on the page.